keyboards/oddball/oddball.c

```diff
@@ -15,3 +15,144 @@
  */
 
 #include "oddball.h"
+#include "pointing_device.h"
+#include "optical_sensor/optical_sensor.h"
+
+#define CLAMP_HID(value) value < -127 ? -127 : value > 127 ? 127 : value
+
+static bool scroll_pressed;
+static int8_t scroll_h;
+static int8_t scroll_v;
+
+void on_cpi_button(int16_t cpi) {
+
+    // read cpi first to prevent unnecessary writes to EEPROM
+    if(optical_sensor_get_config().cpi == cpi)
+        return;
+
+    optical_sensor_set_config((config_optical_sensor_t){ cpi });
+
+    config_oddball_t kb_config;
+    kb_config.cpi = cpi;
+    eeconfig_update_kb(kb_config.raw);
+}
+
+void pointing_device_init(void){
+    if(!is_keyboard_master())
+        return;
+
+    optical_sensor_init();
+
+    // read config from EEPROM and update if needed
+
+    config_oddball_t kb_config;
+    kb_config.raw = eeconfig_read_kb();
+
+    if(!kb_config.cpi) {
+        kb_config.cpi = CPI_2;
+        eeconfig_update_kb(kb_config.raw);
+    }
+
+    optical_sensor_set_config((config_optical_sensor_t){ kb_config.cpi });
+}
+
+void pointing_device_task(void){
+    if(!is_keyboard_master())
+        return;
+
+    report_mouse_t mouse_report = pointing_device_get_report();
+    report_optical_sensor_t sensor_report = optical_sensor_get_report();
+
+    int8_t clamped_x = CLAMP_HID(sensor_report.x);
+    int8_t clamped_y = CLAMP_HID(sensor_report.y);
+
+    if(scroll_pressed) {
+
+        // accumulate scroll
+        scroll_h += clamped_x;
+        scroll_v += clamped_y;
+
+        int8_t scaled_scroll_h = scroll_h / SCROLL_DIVIDER;
+        int8_t scaled_scroll_v = scroll_v / SCROLL_DIVIDER;
+
+        // clear accumulated scroll on assignment
+
+        if(scaled_scroll_h != 0){
+            mouse_report.h = -scaled_scroll_h;
+            scroll_h = 0;
+        }
+
+        if(scaled_scroll_v != 0){
+            mouse_report.v = -scaled_scroll_v;
+            scroll_v = 0;
+        }
+    }
+    else {
+        mouse_report.x = -clamped_x;
+        mouse_report.y = clamped_y;
+    }
+
+    pointing_device_set_report(mouse_report);
+    pointing_device_send();
```

Comment:
Just a heads up, if you don't check to see if the report has changed, and just blindly send it here, it will keep the system (host) awake, as it's considered movement, even if there is none. 

It's an issue I ran into with the ploopyco trackball and the pimironi trackpoint module.  

However, I think it may be worth adding this sort of check to the core code, actually

Reply:
Thanks for the heads up. That would explain why my PC never sleeps; lucky I don't use a laptop!

I've added a check to address this.